internal/pkg/agent/application/upgrade/step_download.go

```diff
@@ -119,3 +118,72 @@ func newVerifier(version string, log *logger.Logger, settings *artifact.Config)
 
 	return composed.NewVerifier(fsVerifier, snapshotVerifier, remoteVerifier), nil
 }
+
+func (u *Upgrader) downloadWithRetries(
```

Comment:
Looking at this I'm not sure we shouldn't just replace it with the package I suggested earlier. It does everything in this function out of the box and is guaranteed to work. I would rather have less complexity to maintain than only one backoff implementation.

Here's an example to replace this entire function:


```go
package main

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/cenkalti/backoff/v4"
)

func main() {
	// This test should only print 3 retries, one per second until the context timeout is hit.
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	boMaxRetries := backoff.WithMaxRetries(backoff.NewConstantBackOff(time.Second), 6)
	boCtx := backoff.WithContext(boMaxRetries, ctx)

	backoffAttempt := 0
	backoff.RetryNotify(
		func() error {
            // This is where we would attempt the download
			backoffAttempt++
			fmt.Printf("attempt: %d\n", backoffAttempt)
			return errors.New("always fails")
		},
		boCtx,
		func(err error, retryAfter time.Duration) {
            // This is a notification callback
			fmt.Printf("Failed err: %s retryAfter: %s\n", err, retryAfter)
		},
	)
}
```

Reply:
I would actually be in favour of replacing our existing backoff implementation with this one everywhere, if we did only want one.